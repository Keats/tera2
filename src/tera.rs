use crate::errors::{Error, TeraResult};
use crate::template::{find_parents, Template};
use crate::vm::interpreter::VirtualMachine;
use crate::{escape_html, Context, HashMap};

/// Default template name used for `Tera::render_str` and `Tera::one_off`.
const ONE_OFF_TEMPLATE_NAME: &str = "__tera_one_off";

/// The escape function type definition
pub type EscapeFn = fn(&str) -> String;

#[derive(Clone)]
pub struct Tera {
    #[doc(hidden)]
    pub templates: HashMap<String, Template>,
    /// Which extensions does Tera automatically autoescape on.
    /// Defaults to [".html", ".htm", ".xml"]
    #[doc(hidden)]
    autoescape_suffixes: Vec<&'static str>,
    #[doc(hidden)]
    escape_fn: EscapeFn,
}

impl Tera {
    pub fn new() -> Self {
        Self::default()
    }

    /// Select which suffix(es) to automatically do HTML escaping on.
    ///
    /// By default, autoescaping is performed on `.html`, `.htm` and `.xml` template files. Only
    /// call this function if you wish to change the defaults.
    ///
    /// # Examples
    ///
    /// Basic usage:
    ///
    /// ```
    /// # use tera::Tera;
    /// let mut tera = Tera::default();
    /// // escape only files ending with `.php.html`
    /// tera.autoescape_on(vec![".php.html"]);
    /// // disable autoescaping completely
    /// tera.autoescape_on(vec![]);
    /// ```
    pub fn autoescape_on(&mut self, suffixes: Vec<&'static str>) {
        self.autoescape_suffixes = suffixes;
    }
    /// Set user-defined function that is used to escape content.
    ///
    /// Often times, arbitrary data needs to be injected into a template without allowing injection
    /// attacks. For this reason, typically escaping is performed on all input. By default, the
    /// escaping function will produce HTML escapes, but it can be overridden to produce escapes
    /// more appropriate to the language being used.
    ///
    /// Inside templates, escaping can be turned off for specific content using the `safe` filter.
    /// For example, the string `{{ data }}` inside a template will escape data, while `{{ data |
    /// safe }}` will not.
    ///
    /// # Examples
    ///
    /// Basic usage:
    ///
    /// ```
    /// # use tera::{Tera, Context};
    /// // Create new Tera instance
    /// let mut tera = Tera::default();
    ///
    /// // Override escape function
    /// tera.set_escape_fn(|input| {
    ///     input.escape_default().collect()
    /// });
    ///
    /// // Create template and enable autoescape
    /// tera.add_raw_template("hello.js", "const data = \"{{ content }}\";").unwrap();
    /// tera.autoescape_on(vec!["js"]);
    ///
    /// // Create context with some data
    /// let mut context = Context::new();
    /// context.insert("content", &"Hello\n\'world\"!");
    ///
    /// // Render template
    /// let result = tera.render("hello.js", &context).unwrap();
    /// assert_eq!(result, r#"const data = "Hello\n\'world\"!";"#);
    /// ```
    pub fn set_escape_fn(&mut self, function: EscapeFn) {
        self.escape_fn = function;
    }

    /// Reset escape function to default [`escape_html()`].
    pub fn reset_escape_fn(&mut self) {
        self.escape_fn = escape_html;
    }

    /// Optimizes the templates when possible and doing some light
    /// checks like whether blocks/macros/templates all exist when they are used
    fn finalize_templates(&mut self) -> TeraResult<()> {
        // {file -> {name -> compiled macro}}
        let mut compiled_macros_by_file = HashMap::with_capacity(20);
        let mut tpl_parents = HashMap::new();
        let mut tpl_size_hint = HashMap::new();

        // 1st loop: we find the parents of each templates with all the block definitions
        // as well as copying the macro definitions defined in each template
        for (name, tpl) in &self.templates {
            let parents = find_parents(&self.templates, tpl, tpl, vec![])?;
            compiled_macros_by_file.insert(tpl.name.clone(), tpl.macro_definitions.clone());

            let mut size_hint = tpl.raw_content_num_bytes;
            for parent in &parents {
                size_hint += &self.templates[parent].raw_content_num_bytes;
            }

            tpl_parents.insert(name.clone(), parents);
            tpl_size_hint.insert(name.clone(), size_hint);
        }

        // 2nd loop: we find the macro definition for each macro calls in each template
        // {tpl_name: vec![compiled macro]|
        let mut tpl_macro_definitions = HashMap::with_capacity(self.templates.len());
        let mut tpl_blocks = HashMap::with_capacity(self.templates.len());
        for (name, tpl) in &self.templates {
            let mut definitions = Vec::new();
            for (tpl_name, macro_name) in &tpl.macro_calls {
                let tpl_w_definition = self.templates.get(tpl_name).ok_or_else(|| {
                    Error::message(format!(
                        "Template `{name}` loads macros from `{tpl_name}` which isn't present in Tera"
                    ))
                })?;

                let definition = tpl_w_definition
                    .macro_definitions
                    .get(macro_name)
                    .ok_or_else(|| {
                        Error::message(format!(
                            "Template `{name}` is using macros {macro_name} from `{tpl_name}` which wasn't found",
                        ))
                    })?;

                definitions.push(definition.clone());
            }
            tpl_macro_definitions.insert(name.to_string(), definitions);

            // TODO: can we avoid cloning the chunk?
            // TODO: Can we just point to things while avoiding hashmap lookups
            let mut blocks = HashMap::with_capacity(tpl.blocks.len());
            for (block_name, chunk) in &tpl.blocks {
                let mut all_blocks = vec![chunk.clone()];
                if chunk.is_calling_function("super") {
                    for parent_tpl_name in tpl_parents[name].iter().rev() {
                        let parent_tpl = self.get_template(parent_tpl_name)?;
                        if let Some(parent_chunk) = parent_tpl.blocks.get(block_name) {
                            all_blocks.push(parent_chunk.clone());
                            if !parent_chunk.is_calling_function("super") {
                                break;
                            }
                        }
                    }
                }
                blocks.insert(block_name.clone(), all_blocks);
            }
            tpl_blocks.insert(name.clone(), blocks);
        }

        // 3rd loop: we actually set everything we've done on the templates object
        for (name, tpl) in self.templates.iter_mut() {
            tpl.raw_content_num_bytes += tpl_size_hint.remove(name.as_str()).unwrap();
            tpl.parents = tpl_parents.remove(name.as_str()).unwrap();
            tpl.macro_calls_def = tpl_macro_definitions.remove(name.as_str()).unwrap();
            tpl.block_lineage = tpl_blocks.remove(name.as_str()).unwrap();
        }

        Ok(())
    }

    /// Add a single template to the Tera instance.
    ///
    /// This will error if the inheritance chain can't be built, such as adding a child
    /// template without the parent one.
    ///
    /// # Bulk loading
    ///
    /// If you want to add several templates, use
    /// [`add_raw_templates()`](Tera::add_raw_templates).
    ///
    /// # Examples
    ///
    /// Basic usage:
    ///
    /// ```
    /// # use tera::Tera;
    /// let mut tera = Tera::default();
    /// tera.add_raw_template("new.html", "Blabla").unwrap();
    /// ```
    pub fn add_raw_template(&mut self, name: &str, content: &str) -> TeraResult<()> {
        let template = Template::new(name, content, None)
            // TODO: need to format the error message with source context
            .map_err(|e| Error::chain(format!("Failed to parse '{}'", name), e))?;

        self.templates.insert(name.to_string(), template);
        self.finalize_templates()?;
        Ok(())
    }

    /// Add all the templates given to the Tera instance
    ///
    /// This will error if the inheritance chain can't be built, such as adding a child
    /// template without the parent one.
    ///
    /// ```no_compile
    /// tera.add_raw_templates(vec![
    ///     ("new.html", "blabla"),
    ///     ("new2.html", "hello"),
    /// ]);
    /// ```
    pub fn add_raw_templates<I, N, C>(&mut self, templates: I) -> TeraResult<()>
    where
        I: IntoIterator<Item = (N, C)>,
        N: AsRef<str>,
        C: AsRef<str>,
    {
        for (name, content) in templates {
            let template = Template::new(name.as_ref(), content.as_ref(), None)
                // TODO: need to format the error message with source context
                .map_err(|e| Error::chain(format!("Failed to parse '{}'", name.as_ref()), e))?;

            self.templates.insert(name.as_ref().to_string(), template);
        }

        self.finalize_templates()?;

        Ok(())
    }

    #[inline]
    pub(crate) fn get_template(&self, template_name: &str) -> TeraResult<&Template> {
        match self.templates.get(template_name) {
            Some(tpl) => Ok(tpl),
            None => Err(Error::template_not_found(template_name)),
        }
    }

    /// Renders a Tera template given a [`Context`].
    ///
    /// # Examples
    ///
    /// Basic usage:
    ///
    /// ```
    /// # use tera::{Tera, Context};
    /// // Create new tera instance with sample template
    /// let mut tera = Tera::default();
    /// tera.add_raw_template("info", "My age is {{ age }}.");
    ///
    /// // Create new context
    /// let mut context = Context::new();
    /// context.insert("age", &18);
    ///
    /// // Render template using the context
    /// let output = tera.render("info", &context).unwrap();
    /// assert_eq!(output, "My age is 18.");
    /// ```
    ///
    /// To render a template with an empty context, simply pass an empty [`Context`] object.
    ///
    /// ```
    /// # use tera::{Tera, Context};
    /// // Create new tera instance with demo template
    /// let mut tera = Tera::default();
    /// tera.add_raw_template("hello.html", "<h1>Hello</h1>");
    ///
    /// // Render a template with an empty context
    /// let output = tera.render("hello.html", &Context::new()).unwrap();
    /// assert_eq!(output, "<h1>Hello</h1>");
    /// ```
    pub fn render(&self, template_name: &str, context: &Context) -> TeraResult<String> {
        let template = self.get_template(template_name)?;
        let mut vm = VirtualMachine::new(self, template);
        vm.render(context)
    }

    /// Renders a one off template (for example a template coming from a user
    /// input) given a `Context` and an instance of Tera. This allows you to
    /// render templates using custom filters or functions.
    ///
    /// Any errors will mention the `__tera_one_off` template: this is the name
    /// given to the template by Tera.
    ///
    /// ```no_compile
    /// let mut context = Context::new();
    /// context.insert("greeting", &"Hello");
    /// let string = tera.render_str("{{ greeting }} World!", &context)?;
    /// assert_eq!(string, "Hello World!");
    /// ```
    pub fn render_str(&mut self, input: &str, context: &Context) -> TeraResult<String> {
        self.add_raw_template(ONE_OFF_TEMPLATE_NAME, input)?;
        let result = self.render(ONE_OFF_TEMPLATE_NAME, context);
        self.templates.remove(ONE_OFF_TEMPLATE_NAME);
        result
    }

    /// Renders a one off template (for example a template coming from a user input) given a `Context`
    ///
    /// This creates a separate instance of Tera with no possibilities of adding custom filters
    /// or testers, parses the template and render it immediately.
    /// Any errors will mention the `__tera_one_off` template: this is the name given to the template by
    /// Tera
    ///
    /// ```no_compile
    /// let mut context = Context::new();
    /// context.insert("greeting", &"hello");
    /// Tera::one_off("{{ greeting }} world", &context, true);
    /// ```
    pub fn one_off(input: &str, context: &Context, autoescape: bool) -> TeraResult<String> {
        let mut tera = Tera::default();

        if autoescape {
            tera.autoescape_on(vec![ONE_OFF_TEMPLATE_NAME]);
        }

        tera.render_str(input, context)
    }
}

impl Default for Tera {
    fn default() -> Self {
        Self {
            templates: HashMap::new(),
            autoescape_suffixes: vec![".html", ".htm", ".xml"],
            escape_fn: escape_html,
        }
    }
}
