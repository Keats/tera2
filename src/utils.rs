use std::fmt;
use std::ops::{Deref, Range};

#[derive(Clone, PartialEq)]
pub struct Spanned<T: fmt::Debug> {
    node: Box<T>,
    span: Span,
}

impl<T: fmt::Debug> Spanned<T> {
    pub fn new(node: T, span: Span) -> Self {
        Self {
            node: Box::new(node),
            span,
        }
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn span_mut(&mut self) -> &mut Span {
        &mut self.span
    }

    pub(crate) fn node(&self) -> &T {
        &self.node
    }

    pub fn into_parts(self) -> (T, Span) {
        (*self.node, self.span)
    }
}

impl<T: fmt::Debug> Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.node
    }
}

impl<T: fmt::Debug> fmt::Debug for Spanned<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.node, f)?;
        write!(f, "{:?}", self.span)
    }
}

#[derive(Clone, PartialEq, Eq, Default)]
pub struct Span {
    pub start_line: usize,
    pub start_col: usize,
    pub end_line: usize,
    pub end_col: usize,
    pub range: Range<usize>,
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            " @ {}:{}-{}:{} ({:?})",
            self.start_line, self.start_col, self.end_line, self.end_col, self.range,
        )
    }
}

impl fmt::Display for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            " @ {}:{}-{}:{}",
            self.start_line, self.start_col, self.end_line, self.end_col,
        )
    }
}
impl Span {
    pub fn expand(&mut self, other: &Span) {
        self.end_line = other.end_line;
        self.end_col = other.end_col;
        self.range = self.range.start..other.range.end;
    }
}

/// Escape HTML following [OWASP](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet)
///
/// Escape the following characters with HTML entity encoding to prevent switching
/// into any execution context, such as script, style, or event handlers. Using
/// hex entities is recommended in the spec. In addition to the 5 characters
/// significant in XML (&, <, >, ", '), the forward slash is included as it helps
/// to end an HTML entity.
///
/// ```text
/// & --> &amp;
/// < --> &lt;
/// > --> &gt;
/// " --> &quot;
/// ' --> &#39;
/// ```
#[inline]
pub fn escape_html(input: &str) -> String {
    let mut output = String::with_capacity(input.len() * 2);
    for c in input.chars() {
        match c {
            '&' => output.push_str("&amp;"),
            '<' => output.push_str("&lt;"),
            '>' => output.push_str("&gt;"),
            '"' => output.push_str("&quot;"),
            '\'' => output.push_str("&#39;"),
            _ => output.push(c),
        }
    }

    // Not using shrink_to_fit() on purpose
    output
}
