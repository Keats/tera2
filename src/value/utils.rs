use serde::ser;
use std::fmt;

#[derive(Debug)]
pub struct SerializationFailed;
impl fmt::Display for SerializationFailed {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "serialization failed")
    }
}
impl std::error::Error for SerializationFailed {}
impl ser::Error for SerializationFailed {
    #[track_caller]
    fn custom<T>(msg: T) -> Self
    where
        T: fmt::Display,
    {
        panic!("{}", msg)
    }
}
