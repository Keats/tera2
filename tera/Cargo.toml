[package]
name = "tera"
version = "2.0.0"
edition = "2021"

[dependencies]
serde = "1"

# Optional deps
indexmap = { version = "2", optional = true }
unic-segment = { version = "0.9", optional = true }
walkdir = { version = "2.5.0", optional = true }
globset = { version = "0.4.14", optional = true }

# Crates for speedup
ahash = { version = "0.8", optional = true }
itoa = { version = "1.0.9", optional = true }
pulldown-cmark-escape = { version = "0.10", optional = true}

[features]
default = []
# Use hashbrown with ahash hashing to improve performance
fast_hash = ["dep:ahash", "ahash?/serde"]
# Keep order of insertion for Value
preserve_order = ["dep:indexmap"]
# If we want Tera to work with graphemes clusters rather than utf-8 characters when iterating on strings
unicode = ["dep:unic-segment"]
# Use itoa to format numbers (instead of std::fmt)
no_fmt = ["dep:itoa"]
# Use a faster HTML escape than the default one
fast_escape = ["dep:pulldown-cmark-escape"]
# Improve performance by adding extra dependencies
fast = ["no_fmt", "fast_escape", "fast_hash"]
# Loads files disks using a glob
glob_fs = ["dep:walkdir", "dep:globset"]

[dev-dependencies]
criterion = "0.5"
iai-callgrind = "0.8.0"
insta = { version = "1", features = ["glob"] }
serde_derive = "1.0.156"
tempfile = "3"

[lib]
doctest = false

[[bench]]
name = "value"
harness = false

[[bench]]
name = "escape"
harness = false

[[bench]]
name = "templates"
harness = false

# you can enable the benches below on Linux provided you have everything installed.
# see https://github.com/iai-callgrind/iai-callgrind for details
#[[bench]]
#name = "iai-value"
#harness = false
#
#[[bench]]
#name = "iai-escape"
#harness = false
#
#[[bench]]
#name = "iai-templates"
#harness = false
