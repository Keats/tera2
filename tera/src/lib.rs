mod args;
mod context;
mod errors;
mod filters;
mod parsing;
mod reporting;
mod template;
mod tera;
mod tests;
mod functions;
mod utils;
pub mod value;
pub(crate) mod vm;

pub use crate::tera::{EscapeFn, Tera};
pub use context::Context;
pub use parsing::parser::Parser;
pub use utils::escape_html;
pub use value::Value;

#[cfg(ahash)]
pub(crate) use hashbrown::{HashMap, HashSet};
#[cfg(not(ahash))]
pub(crate) use std::collections::{HashMap, HashSet};

#[cfg(test)]
mod snapshot_tests;
