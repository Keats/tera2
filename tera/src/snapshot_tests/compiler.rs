use crate::parsing::parser::Parser;
use crate::parsing::Compiler;

#[test]
fn compiler_ok() {
    insta::glob!("compiler_inputs/success/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let nodes = Parser::new(&contents).parse().unwrap().nodes;
        let mut compiler = Compiler::new(&path.file_name().unwrap().to_string_lossy(), "");
        compiler.compile(nodes);

        insta::assert_debug_snapshot!(compiler.chunk);
    });
}

#[test]
fn compiler_blocks() {
    insta::glob!("compiler_inputs/blocks/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let nodes = Parser::new(&contents).parse().unwrap().nodes;
        let mut compiler = Compiler::new(&path.file_name().unwrap().to_string_lossy(), "");
        compiler.compile(nodes);

        let mut s = String::with_capacity(1000);
        s.push_str(&format!("{:?}", compiler.chunk));
        s.push_str("\n\n");

        let mut blocks: Vec<_> = compiler.blocks.into_iter().collect();
        blocks.sort_by(|a, b| a.0.cmp(&b.0));
        for (name, chunk) in blocks {
            s.push_str(&format!(">> Block: {name}\n"));
            s.push_str(&format!("{chunk:?}"));
            s.push_str("---\n");
        }
        insta::assert_snapshot!(s);
    });
}
