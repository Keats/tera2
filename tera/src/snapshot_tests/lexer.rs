use crate::parsing::lexer::tokenize;

#[test]
fn lexer_ok() {
    insta::glob!("lexer_inputs/success/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let tokens: Result<Vec<_>, _> = tokenize(&contents).collect();
        let tokens = tokens.unwrap().into_iter().map(|x| x.0).collect::<Vec<_>>();
        insta::assert_debug_snapshot!(&tokens);
    });
}

#[test]
fn lexer_errors() {
    insta::glob!("lexer_inputs/errors/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let res: Result<Vec<_>, _> = tokenize(&contents).collect();
        assert!(res.is_err());
        insta::assert_debug_snapshot!(res.unwrap_err());
    });
}
