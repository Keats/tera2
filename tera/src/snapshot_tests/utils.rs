use crate::Tera;

/// This will take our custom insta format to create multiple templates
/// The format is:
/// $$ filename
/// body
/// $$ other filename
/// other body
pub fn split_multi_templates(body: &str) -> Vec<(String, String)> {
    let parts: Vec<_> = body.split("$$ ").skip(1).collect();
    let mut tpls = Vec::with_capacity(parts.len());
    for part in parts {
        let mut chars = part.chars();
        let filename: String = chars.by_ref().take_while(|&c| c != '\n').collect();
        let content = chars.collect::<String>().trim().to_string();
        tpls.push((filename, content));
    }
    tpls
}

/// Normalizes line endings by converting Windows CRLF (\r\n) to Unix LF (\n).
/// This ensures cross-platform compatibility for snapshot tests.
/// Only used in test code - no performance impact on production.
pub fn normalize_line_endings(content: &str) -> String {
    content.replace("\r\n", "\n")
}

/// Splits the body into multiple templates ans
/// returns the tera instance as well as the last template name.
/// To be used when the templates are valid only
pub fn create_multi_templates_tera(body: &str) -> (Tera, String) {
    let normalized_body = normalize_line_endings(body);
    let tpls = split_multi_templates(&normalized_body);
    let last_filename = tpls.last().unwrap().0.clone();
    let mut tera = Tera::default();
    tera.add_raw_templates(tpls).unwrap();

    (tera, last_filename)
}
