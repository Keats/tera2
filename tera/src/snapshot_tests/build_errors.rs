//! Build errors are errors that Tera catches when loading templates
//! eg a template inheriting from a template that isn't there
use crate::tera::Tera;

use crate::snapshot_tests::utils::{normalize_line_endings, split_multi_templates};

#[test]
fn build_errors() {
    insta::glob!("build_errors/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let normalized_contents = normalize_line_endings(&contents);
        let tpls = split_multi_templates(&normalized_contents);
        let mut tera = Tera::default();
        let err = tera.add_raw_templates(tpls).unwrap_err();
        insta::assert_snapshot!(&err);
    });
}
