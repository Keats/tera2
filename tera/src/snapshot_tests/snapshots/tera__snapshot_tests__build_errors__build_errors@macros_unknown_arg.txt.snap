---
source: tera/src/snapshot_tests/build_errors.rs
expression: "&err"
input_file: tera/src/snapshot_tests/build_errors/macros_unknown_arg.txt
---
Template tpl is calling macro hello with an argument lang which isn't present in its definition. Only the following are allowed: other, val.
